### PR TITLE
internal/secrets: add Google Cloud KMS implementation

### PR DESCRIPTION
**Pull request by shantuo**
Update #420

### CODE REVIEW

File: internal/secrets/gcpkms/kms_test.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+func newHarness(ctx context.Context, t *testing.T) (drivertest.Harness, error) {
+	conn, done := setup.NewGCPgRPCConn(ctx, t, endPoint)
+	client, err := cloudkms.NewKeyManagementClient(ctx, option.WithGRPCConn(conn))
+	if err != nil {
+		return nil, err
+	}
+	return &harness{
+		client: client,
+		close: func() {
+			_ = client.Close()
```
Comment (clausti) [inline internal/secrets/gcpkms/kms_test.go]: Why does `client.Close()` get assigned to anything here? (I am assuming there's a reason it needs to be, just curious what it is.)

File: internal/secrets/gcpkms/example_test.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+package gcpkms_test
+
+import (
+	"context"
+
+	"github.com/google/go-cloud/internal/secrets/gcpkms"
+)
+
+func ExampleCrypter_Encrypt() {
+	ctx := context.Background()
```
Comment (zombiezen) [inline internal/secrets/gcpkms/example_test.go]: For here and the other example: it's helpful to over-explain each step in examples so that the reader knows what's going and on, and especially what they need to change.

File: internal/secrets/gcpkms/kms.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+// DefaultClient returns a client to use with Cloud KMS and a clean-up function
+// to close the client after used.
+func DefaultClient(ctx context.Context, ts gcp.TokenSource) (*cloudkms.KeyManagementClient, func(), error) {
+	c, err := cloudkms.NewKeyManagementClient(ctx, option.WithTokenSource(ts))
+	return c, func() { c.Close() }, err
+}
+
+// NewCrypter returns a new Crypter to to encryption and decryption.
+func NewCrypter(ki *KeyInfo, client *cloudkms.KeyManagementClient) *Crypter {
```
Comment (zombiezen) [inline internal/secrets/gcpkms/kms.go]: nit: Flip the order of these parameters. Go from most general type to least general.

```suggestion
func NewCrypter(client *cloudkms.KeyManagementClient, ki *KeyInfo) *Crypter {
```

File: internal/secrets/gcpkms/kms.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	"google.golang.org/api/option"
+	kmspb "google.golang.org/genproto/googleapis/cloud/kms/v1"
+)
+
+// endPoint is the address to access Google Cloud KMS API.
+const endPoint = "cloudkms.googleapis.com:443"
+
+// DefaultClient returns a client to use with Cloud KMS and a clean-up function
+// to close the client after used.
+func DefaultClient(ctx context.Context, ts gcp.TokenSource) (*cloudkms.KeyManagementClient, func(), error) {
```
Comment (zombiezen) [inline internal/secrets/gcpkms/kms.go]: nit: `Dial` for consistency with other packages?

File: internal/secrets/gcpkms/kms.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+func NewCrypter(ki *KeyInfo, client *cloudkms.KeyManagementClient) *Crypter {
+	return &Crypter{
+		keyInfo: ki,
+		client:  client,
+	}
+}
+
+// KeyInfo includes related information to construct a key name that is managed
+// by Cloud KMS.
+type KeyInfo struct {
```
Comment (zombiezen) [inline internal/secrets/gcpkms/kms.go]: This concept is called the [key resource ID](https://cloud.google.com/kms/docs/object-hierarchy#key) in the docs. So perhaps `KeyID` is more suitable here, as well as a link to that doc?

File: internal/secrets/gcpkms/kms.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+	ProjectID, Location, KeyRing, KeyID string
+}
+
+// Crypter contains information to construct the pull path of a key.
+type Crypter struct {
+	keyInfo *KeyInfo
+	client  *cloudkms.KeyManagementClient
+}
+
+func (ki *KeyInfo) keyPath() string {
```
Comment (zombiezen) [inline internal/secrets/gcpkms/kms.go]: Move this to be next to the `KeyInfo` type. Also seems reasonable to make this the `String()` method for the type.

File: internal/secrets/gcpkms/kms.go [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+// KeyInfo includes related information to construct a key name that is managed
+// by Cloud KMS.
+type KeyInfo struct {
+	ProjectID, Location, KeyRing, KeyID string
+}
+
+// Crypter contains information to construct the pull path of a key.
+type Crypter struct {
```
Comment (zombiezen) [inline internal/secrets/gcpkms/kms.go]: Since secrets is going to use the driver pattern, this type should not be exported.

File: internal/secrets/gcpkms/kms_test.go [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+	"github.com/google/go-cloud/internal/testing/setup"
+	"google.golang.org/api/option"
+
+	cloudkms "cloud.google.com/go/kms/apiv1"
+	"github.com/google/go-cloud/internal/secrets/driver"
+	"github.com/google/go-cloud/internal/secrets/drivertest"
+)
+
+const (
+	// These constants capture values that were used during the last --record.
```
Comment (zombiezen) [inline internal/secrets/gcpkms/kms_test.go]: Move the doc comment to right before the `const` keyword. Otherwise, the doc is just associated with `projectID`.

File: runtimevar/runtimevar.go [hunk 9]
```diff
@@ -122,10 +122,11 @@ func (w *wrappedError) Error() string {
 // Decode is a function type for unmarshaling/decoding bytes into given object.
 type Decode func([]byte, interface{}) error
 
-// Decoder is a helper for decoding bytes into a particular Go type object.  The Variable objects
-// produced by a particular driver.Watcher should always contain the same type for Variable.Value
-// field.  A driver.Watcher can use/construct a Decoder object with an associated type (Type) and
-// decoding function (Func) for decoding retrieved bytes into Variable.Value.
+// Decoder is a helper for decoding bytes into a particular Go type object. The
+// Variable objects produced by a particular driver.Watcher should always
```
Comment (zombiezen) [inline runtimevar/runtimevar.go]: Unrelated change. Revert?

File: internal/secrets/gcpkms/example_test.go (hunk 2 above)
Comment (shantuo) [reply internal/secrets/gcpkms/example_test.go]: Added some comments, thanks!

**Comment by shantuo** [comment]
PTAL
